[FLINK-32409][runtime] Makes DefaultLeaderElectionService not rely on MultipleComponentLeaderElectionDriverAdapter anymore

XComp opened the pull request:
## PR Hierarchy

* https://github.com/apache/flink/pull/21742
* https://github.com/apache/flink/pull/22379
* https://github.com/apache/flink/pull/22422
* https://github.com/apache/flink/pull/22380
* https://github.com/apache/flink/pull/22623
* https://github.com/apache/flink/pull/22384
* https://github.com/apache/flink/pull/22390
* https://github.com/apache/flink/pull/22404
* https://github.com/apache/flink/pull/22601
* https://github.com/apache/flink/pull/22642
* https://github.com/apache/flink/pull/22640
* https://github.com/apache/flink/pull/22656
* https://github.com/apache/flink/pull/22828
* https://github.com/apache/flink/pull/22830
* https://github.com/apache/flink/pull/22829
* https://github.com/apache/flink/pull/22661
* https://github.com/apache/flink/pull/22844
* === THIS PR === FLINK-32409
* https://github.com/apache/flink/pull/22873
* https://github.com/apache/flink/pull/22935

## What is the purpose of the change

The final cleanup PR which removes the dependency between `MultipleComponentLeaderElectionDriverAdapter` and `DefaultLeaderElectionService`.

## Brief change log

* Utilization of the multi-component support of the `DefaultLeaderElectionService` implementation
* Removal of the adapter functionality
* Removal of any unused classes (that might be moved into its own PR to reduce the diff)

## Verifying this change

* legacy tests were moved one by one into the test classes that are kept

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java:
```diff
@@ -242,28 +251,15 @@ public CompletableFuture<Void> globalCleanupAsync(JobID jobID, Executor executor
-        return leaderElectionService.createLeaderElection("unused-contender-id");
-    }
+        synchronized (lock) {
+            if (leaderElectionService == null) {
```
We have to start the leader election lazily here. Otherwise, `JobManagerHAProcessFailureRecoveryITCase` would stall. It looks like the reason is that we create `TaskManagers` sequentially there which rely on the `ResourceManager` being around (for setting up the Rpc bind address).

Creating the driver in the `AbstractHaServices` constructor will trigger the leadership grant earlier without the `ResourceManager` being instantiated, yet. This makes the address resolution fail and the test won't succeed.

Therefore, I kept it like that for now (because that's how the `DefaultMultipleComponentLeaderElectionService` was instantiated as well). But I still have to investigate whether it's actually true and whether we can see the same behavior in `release-1.17`.

XComp commented:
Force-pushed a rebase to the most-recent version of FLINK-32436 to include master with the fix for FLINK-32478

XComp commented:
Force-pushed a rebase to the most-recent version of [FLINK-32436](https://issues.apache.org/jira/browse/FLINK-32436) to include master with the fix for FLINK-32490

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java:
```diff
@@ -161,7 +166,9 @@ public void close() throws Exception {
 
         try {
-            closeableRegistry.close();
+            if (leaderElectionService != null) {
+                leaderElectionService.close();
+            }
```
shouldnt this happen under the lock?

XComp replied:
Correct. :+1:

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java:
```diff
@@ -70,10 +68,15 @@ public abstract class AbstractHaServices implements HighAvailabilityServices {
+
+    private final Object lock = new Object();
+
+    private DefaultLeaderElectionService leaderElectionService;
```
add GuardedBy("lock")?

zentol commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -69,75 +59,86 @@ public class KubernetesMultipleComponentLeaderElectionHaServices extends Abstrac
+                blobStoreService);
+    }
 
+    private KubernetesMultipleComponentLeaderElectionHaServices(
```
not sure why we need 2 private constructors.

XComp replied:
Fair enough. I'm gonna remove the intermediate one. Initially, I created it because I wanted to avoid having `configuration.get(KubernetesConfigOptions.CLUSTER_ID)` being called twice for the `clusterId`.

XComp commented:
force-pushed a rebase to `master` after FLINK-31784 was merged.

zentol commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -69,75 +59,72 @@ public class KubernetesMultipleComponentLeaderElectionHaServices extends Abstrac
+        this.kubeClient = checkNotNull(kubeClient);
+        this.clusterId = checkNotNull(clusterId);
+        this.configMapSharedWatcher = checkNotNull(configMapSharedWatcher);
+        this.watchExecutorService = checkNotNull(watchExecutorService);
```
Where is this executor closed?

XComp commented:
I added the hotfix for the shutdown logic and squashed the commits together. I will do a separate rebase to most-recent (stable) version of `master` (44e1397e2e7aa54667bf3b30bc6f6d5db32b5f79) afterwards to trigger a final CI run.
